File: src/dotnet/Ansight.Tools.Preferences/GetPreferenceValueTool.cs
namespace Ansight.Tools.Preferences;

public sealed class GetPreferenceValueTool : ITool
{
    private readonly PreferencesToolsOptions options;

    public GetPreferenceValueTool(PreferencesToolsOptions? options = null)
    {
        this.options = options ?? PreferencesToolsOptions.Default;
    }

    public string Category => "prefs";

    public ToolPolicy Policy => ToolPolicy.Read;

    public string Id => PreferencesToolIds.GetValue;

    public string Name => "Get Preference Value";

    public string Description => "Reads a value from shared preferences or user defaults.";

    public string Keywords => "preferences sharedpreferences nsuserdefaults storage get";

    public ToolSchema ArgumentsSchema => PreferencesToolSchemas.GetValueArguments;

    public ToolSchema ResultSchema => PreferencesToolSchemas.GetValueResult;

    public Task<ToolResult> Execute(IReadOnlyDictionary<string, string> arguments)
    {
        ArgumentNullException.ThrowIfNull(arguments);

        try
        {
            return Task.FromResult(ToolResult.Success(PreferencesSupport.GetValue(options, arguments)));
        }
        catch (PlatformNotSupportedException exception)
        {
            return Task.FromResult(ToolResult.Failure(exception.Message, errorCode: "prefs_platform_unsupported"));
        }
        catch (Exception exception)
        {
            return Task.FromResult(ToolResult.Failure(exception.Message, errorCode: "prefs_get_failed"));
        }
    }
}

File: src/ios/Sources/AnsightToolsPreferences/GetPreferenceValueTool.swift
import AnsightKit
import Foundation

public final class GetPreferenceValueTool: AnsightTool {
    private let options: AnsightPreferencesToolOptions

    public init(options: AnsightPreferencesToolOptions = .default) {
        self.options = options
    }

    public var descriptor: AnsightToolDescriptor {
        AnsightToolDescriptor(
            id: AnsightPreferencesToolIds.getValue,
            name: "Get Preference Value",
            description: "Reads a value from an iOS UserDefaults store.",
            category: "prefs",
            scope: AnsightToolScope.read.rawValue,
            keywords: "preferences userdefaults read settings",
            security: AnsightPreferencesToolSecurityProfiles.getValue,
            argumentsSchema: AnsightPreferencesToolSchemas.getValueArguments,
            resultSchema: AnsightPreferencesToolSchemas.getValueResult
        )
    }

    public func execute(arguments: [String: String]) throws -> AnsightToolExecutionResult {
        do {
            return .success(try AnsightPreferencesSupport.getValue(options: options, arguments: arguments))
        } catch {
            return .failure(error.localizedDescription, errorCode: "prefs_get_failed")
        }
    }
}
